datalake/tests: test with transactions

bharathv opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv commented:
/dt

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--b46ee388-1c98-4354-8f0a-8592e22c1dff-->
<!--NEW FAILURES-->
<!--fab9171d-67e6-4a1e-a149-7cc526e244d6-->
<!--NEW FAILURES-->
<!--cf1bc76c-f851-401d-bf3b-4c40dfd9212a-->
<!--NEW FAILURES-->
<!--6870a81c-30a3-4c42-b7cb-5cce0ab350e2-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59340#019399cc-77c2-478c-b4fe-e796b1013c72:
```
"rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.S3"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59548#0193b19d-8626-4c53-93e4-ac00f3bc3117:
```
"rptest.tests.datalake.transactions_test.DatalakeTransactionTests.test_with_transactions.cloud_storage_type=CloudStorageType.S3.compaction=False"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59548#0193b19d-8627-404c-820e-3fce1a82f21f:
```
"rptest.tests.datalake.transactions_test.DatalakeTransactionTests.test_with_transactions.cloud_storage_type=CloudStorageType.S3.compaction=False"
```
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59577#0193b35a-be91-41e9-abf8-986d2aa152fe:
```
"rptest.tests.e2e_shadow_indexing_test.EndToEndHydrationTimeoutTest.test_hydration_completes_when_consumer_killed"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59340-->
<!--1c738b9c-d5c9-4879-9724-a70123df5177-->
<!--title-->
## Retry command for [Build#59340](https://buildkite.com/redpanda/redpanda/builds/59340)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
```
<!--/body-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--981cf551-085b-4560-90b2-debf6e2454bb-->
<!--RETRIED FAILURES-->
<!--24e61eb8-1cb7-46d0-a6c7-758713d96ab5-->
<!--RETRIED FAILURES-->
<!--661cfa3d-734b-4b76-846b-fce431ceb119-->
<!--RETRIED FAILURES-->
<!--3659ece9-7e23-4ac4-9af6-e7264c6b8f78-->
<!--RETRIED FAILURES-->
<!--f81985d3-3dba-4292-9f11-d3f5c5d6641a-->
<!--RETRIED FAILURES-->
<!--83d8b685-eb7b-46a9-ad5a-1654f7ad608d-->
<!--RETRIED FAILURES-->
<!--4a44d869-c99b-4dbf-889e-06d9c6edfbca-->
<!--RETRIED FAILURES-->
<!--6ba1c412-1bae-4624-a448-d46ebc433702-->
<!--RETRIED FAILURES-->
<!--17a15c48-967e-4872-8f95-19f841d42b36-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59340#019399cc-77c0-482f-819a-b311c57567a7
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59369#01939b77-6540-4b64-9eaf-45720283a367
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59369#01939b7c-5ac3-445b-a098-53396df74688
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59532#0193b032-e42a-416c-bff2-c1ece3f68ebd
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59577#0193b35a-be92-4870-82f7-ac5267b81790
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59577#0193b35f-ab27-4391-b027-6201ce8081ea
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59577#0193b35f-ab2b-4339-97f8-6402965160c7
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59577#0193b35f-ab2d-41e5-8a80-0dc81be165ad
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59577#0193b35f-ab2e-4a63-ba6b-d8b5db4446be
<!--/body-->

bharathv committed:
datalake/verifier: add ability to handle transactions with gaps

Currently verifier expects to be able to consume/query until hwm. This
does not hold true for transactions due to the offset gaps created by
control batches (commit/batches) and aborted data batches.

A simple example is when the topic data ends with an aborted transaction
resulting in a range of offsets being non consumable.

This commit changes the invariants that mark completion of translation
to the following

1. All offsets until hwm should be consumed
2. All consumed messages must be translated
3. Max consumption offset should be same max translated offset

(1) is where most of the change is as we cannot rely on message offsets
to determine if the consumption is complete. Eg: the topic ends with an
aborted transaction (which includes 1 or more aborted batches and a
control batch that is not propagated to the app layer)

bharathv commented:
/ci-repeat 3
skip-units
dt-repeat=100
tests/rptest/tests/datalake/transactions_test.py

andrwng commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -193,9 +198,13 @@ def _all_offsets_translated(self):
 
-                if self._max_queried_offsets[p.id] < p.high_watermark - 1:
+                max_offset_to_check = p.high_watermark - 1
+                if self._max_offset_check_override:
```
nit: `is not None`, given _max_offset_check_override = 0 also skips this check?

bharathv replied:
removed this code.

andrwng commented on tests/rptest/tests/datalake/transactions_test.py:
```diff
@@ -0,0 +1,139 @@
+            # Issue a last committed transaction so we do not end with an aborted
+            # transaction which makes the verification difficult (as we need to keep
+            # track of last committed transaction offset)
+            offset = self.append_last_committed_transaction()
+            tx_verifier.stop()
```
Just curious, does it matter the order of stopping and appending the last committed transaction?

bharathv replied:
no it doesn't.  the final stop just verifies some invariants on the service side.

andrwng commented on tests/rptest/tests/datalake/transactions_test.py:
```diff
@@ -0,0 +1,139 @@
+            offset = self.append_last_committed_transaction()
+            tx_verifier.stop()
+
+            #dl.wait_for_translation_until_offset(self.topic_name,
+            #                                     offset=offset)
```
nit: remove?

andrwng commented on tests/rptest/tests/datalake/datalake_verifier.py:
```diff
@@ -193,9 +198,13 @@ def _all_offsets_translated(self):
                     )
                     return False
 
-                if self._max_queried_offsets[p.id] < p.high_watermark - 1:
```
Not necessarily from this PR, but wondering if it makes sense to ensure that there are no offsets higher than `_max_offset_check_override` (or some other variable)? IIUC this passes if the table contains offset 1000 but the HWM is 500, which seems possible with transactions

bharathv replied:
I think the check makes sense in general (unrelated to this work) but I ended up revamping the whole thing with a new approach that seems more cleaner.

andrwng reviewed (commented):
Nice

bharathv committed:
datalake/tests: test with transactions

bharathv commented:
/ci-repeat 3
skip-units
dt-repeat=50
tests/rptest/tests/datalake/transactions_test.py

bharathv commented:
/ci-repeat 3
skip-units
dt-repeat=50
tests/rptest/tests/datalake/transactions_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59548-->
<!--c4821d33-3b66-4091-b217-45e3d0b81a2a-->
<!--aaefa334-901d-4cf5-aa92-787396e79b73-->
<!--title-->
## Retry command for [Build#59548](https://buildkite.com/redpanda/redpanda/builds/59548)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/transactions_test.py::DatalakeTransactionTests.test_with_transactions@{"cloud_storage_type":1,"compaction":false}
```
<!--/body-->

bharathv commented:
/ci-repeat 3
skip-units
dt-repeat=100
tests/rptest/tests/datalake/transactions_test.py

bharathv commented:
Ready for review.

Note to the reviewers: had to force push to resolve conflicts. Latest implementation does not rely on forcing a committed transaction as the last transaction, instead the verifier now keeps track of last consumable offset and it tolerant to trailing offset gaps caused by aborted/control batches.

bharathv commented:
/ci-repeat 3
skip-units
tests/rptest/tests/datalake/

andrwng reviewed (approved):
lgtm

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59577-->
<!--00a13df0-85f9-4ed8-ab31-c1f10b4d9ede-->
<!--title-->
## Retry command for [Build#59577](https://buildkite.com/redpanda/redpanda/builds/59577)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndHydrationTimeoutTest.test_hydration_completes_when_consumer_killed
```
<!--/body-->
